File: src/main/java/de/twometer/amongus/net/server/AmongUsServer.java
package de.twometer.amongus.net.server;

import com.esotericsoftware.kryonet.Connection;
import com.esotericsoftware.kryonet.Listener;
import com.esotericsoftware.kryonet.Server;
import de.twometer.amongus.model.*;
import de.twometer.amongus.net.NetMessage;
import de.twometer.amongus.util.AsyncScheduler;
import de.twometer.amongus.util.Config;
import de.twometer.amongus.util.RandomUtil;
import de.twometer.neko.util.Log;
import de.twometer.neko.util.MathF;
import org.joml.Vector3f;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class AmongUsServer extends Listener {

    private final Server server = new Server() {
        @Override
        protected Connection newConnection() {
            return new PlayerConnection();
        }
    };

    private final Map<String, ServerSession> sessions = new ConcurrentHashMap<>();
    private final HandlerRegistry handlers = new HandlerRegistry();
    private final AsyncScheduler scheduler = new AsyncScheduler();
    private final AtomicInteger connectedClients = new AtomicInteger();
    private final AtomicInteger idCounter = new AtomicInteger();
    private final Random random = new Random();

    public AmongUsServer() {
        NetMessage.registerAll(server.getKryo());
        scheduler.start();
    }

    public void launch() throws IOException {
        Log.i("Starting Among Us 3D server...");

        var config = Config.get();

        server.start();

        Log.i("Binding to *:" + config.getServerPort());
        server.addListener(this);
        server.bind(config.getServerPort());

        registerHandlers();
        Log.i("Server is online.");
    }

    @Override
    public void connected(Connection connection) {
        var playerId = newId();
        ((PlayerConnection) connection).player.id = playerId;
        Log.i(connection.getRemoteAddressTCP().toString() + " (#" + playerId + ") connected. " + connectedClients.incrementAndGet() + " clients online.");
    }

    @Override
    public void disconnected(Connection c) {
        var connection = ((PlayerConnection) c);
        if (connection.session != null) {
            connection.session.removePlayer(connection.player.id);
            connection.session.broadcast(new NetMessage.OnPlayerLeave(connection.player.id));

            if (connection.session.getHost() == connection.player.id) {
                Log.i("Host left, finding new host");
                connection.session.findNewHost();
                connection.session.broadcast(new NetMessage.OnHostChanged(connection.session.getHost()));
            }

            if (connection.session.getPlayers().size() == 0) {
                Log.i("Session is empty, deleting it.");
                sessions.remove(connection.session.getGameCode());
            }
        }
        var playerId = connection.getId();
        Log.i("#" + playerId + " disconnected. " + connectedClients.decrementAndGet() + " clients online.");
    }

    @Override
    public void received(Connection c, Object o) {
        var connection = (PlayerConnection) c;
        //Log.d("Incoming message: " + o.toString());
        handlers.handle(connection, o);
    }

    private void registerHandlers() {
        handlers.register(NetMessage.SessionJoin.class, (p, m) -> {
            var session = getSession(m.code);
            if (session == null)
                p.sendTCP(new NetMessage.SessionJoined(NetMessage.SessionJoined.Result.InvalidGameCode));
            else if (m.username == null || m.username.trim().length() == 0)
                p.sendTCP(new NetMessage.SessionJoined(NetMessage.SessionJoined.Result.InvalidUsername));
            else if (session.isUsernameTaken(m.username))
                p.sendTCP(new NetMessage.SessionJoined(NetMessage.SessionJoined.Result.UsernameTaken));
            else if (session.isFull())
                p.sendTCP(new NetMessage.SessionJoined(NetMessage.SessionJoined.Result.LobbyFull));
            else {
                // Configure the player object
                p.session = session;
                p.player.username = m.username;
                p.player.role = PlayerRole.Crewmate;
                p.player.color = p.session.getRandomFreeColor();

                // Send join success message
                p.sendTCP(new NetMessage.SessionJoined(p.player.id, session.getGameCode(), NetMessage.SessionJoined.Result.Success, p.session.getHost()));

                // Send list of currently online players
                for (var player : session.getPlayers())
                    p.sendTCP(new NetMessage.OnPlayerJoin(player.getId(), player.getUsername(), player.getColor()));

                // Send session config
                p.sendTCP(new NetMessage.OnSessionUpdate(session.getConfig()));

                // Broadcast that they joined
                p.session.addPlayer(p);
                p.session.broadcast(new NetMessage.OnPlayerJoin(p.player.id, p.player.username, p.player.color));

                // Init client side player object
                p.sendTCP(new NetMessage.OnPlayerUpdate(p.player.id, p.player.color, p.player.role));
            }
        });
        handlers.register(NetMessage.SessionCreate.class, (p, m) -> {
            var code = CodeGenerator.newGameCode();
            var session = new ServerSession(code, p.player.id);
            sessions.put(code, session);
            p.sendTCP(new NetMessage.SessionCreated(NetMessage.SessionCreated.Result.Ok, code));
        });
        handlers.register(NetMessage.SessionConfigure.class, (p, m) -> {
            if (p.session == null) return;
            if (p.session.getHost() != p.player.id) return;
            p.session.setConfig(m.config);
            p.sendTCP(new NetMessage.SessionConfigured(true));
        });
        handlers.register(NetMessage.ColorChange.class, (p, m) -> {
            if (p.session == null) return;
            if (p.session.isColorAvailable(m.newColor)) {
                p.player.color = m.newColor;
                p.sendTCP(new NetMessage.ColorChanged(true));
                broadcastPlayerUpdate(p);
            } else
                p.sendTCP(new NetMessage.ColorChanged(false));
        });
        handlers.register(NetMessage.StartGame.class, (p, m) -> {
            if (p.session == null) return;
            if (p.session.getHost() != p.player.id) return;

            // Choose impostors
            var playerCount = p.session.getPlayers().size();
            int preferredImpostors = p.session.getConfig().getImpostorCount();
            int maxImpostors = 3;
            if (playerCount < 6) maxImpostors = 1;
            else if (playerCount < 9) maxImpostors = 2;

            var impostors = Math.min(preferredImpostors, maxImpostors);
            for (var i = 0; i < impostors; i++) {
                var imp = randomImpostor(p.session);
                imp.player.role = PlayerRole.Impostor;
                broadcastPlayerUpdate(imp);
            }

            // Assign spawn locations
            tpAllToSpawn(p.session);

            // Assign tasks and start game
            var commonTask = TaskGenerator.newCommonTask();
            for (var player : p.session.getPlayers()) {
                var list = new ArrayList<PlayerTask>();
                list.add(commonTask);
                generateTasks(list, p.session.getConfig());
                Collections.shuffle(list);

                if (player.getRole() != PlayerRole.Impostor)
                    for (var task : list)
                        p.session.totalTaskStages += task.length();

                player.sendTCP(new NetMessage.OnGameStart(list));
            }
        });
        handlers.register(NetMessage.PositionChange.class, (p, m) -> {
            if (p.session == null) return;
            p.player.position = m.position;
            p.player.rotation = m.rotation;
            m.playerId = p.player.id;
            p.session.broadcastExcept(m, m.playerId);
        });
        handlers.register(NetMessage.CompleteTaskStage.class, (p, m) -> {
            if (p.session == null) return;
            p.session.tasksFinished++;
            p.session.broadcast(new NetMessage.OnTaskProgressChanged(p.session.getTaskProgress()));
            checkVictory(p.session);
        });
        handlers.register(NetMessage.CallEmergency.class, (p, m) -> {
            if (p.session == null) return;
            var broadcast = new NetMessage.OnEmergencyMeeting(p.player.id, m.cause);
            p.session.broadcast(broadcast);
            p.session.votingTask = scheduler.runLater(p.session.getConfig().getVotingTime() * 1000, () -> {
                p.session.votingTask = null;
                handleVotingEnd(p.session);
            });
        });
        handlers.register(NetMessage.Vote.class, (p, m) -> {
            if (p.session == null) return;
            if (p.session.votes.containsKey(p.player.id)) // no double votes
                return;

            var broadcast = new NetMessage.OnPlayerVoted(p.player.id, m.playerId);
            p.session.broadcast(broadcast);
            p.session.votes.put(p.player.id, m.playerId);
            checkAllVoted(p.session);
        });
        handlers.register(NetMessage.Kill.class, (p, m) -> {
            if (p.session == null) return;
            var victim = p.session.getPlayer(m.playerId);
            victim.player.alive = false;
            p.session.broadcast(new NetMessage.Kill(false, victim.player.id));
            p.session.broadcast(new NetMessage.PositionChange(p.player.id, victim.player.position, victim.player.rotation));
            checkVictory(p.session);
        });
        handlers.register(NetMessage.StartSabotage.class, (p, m) -> {
            if (p.session == null) return;
            p.session.broadcast(new NetMessage.OnSabotageStateChanged(m.sabotage, true, 30000, CodeGenerator.newO2Code()));
            if (m.sabotage == Sabotage.O2 || m.sabotage == Sabotage.Reactor) {
                p.session.sabotageTask = scheduler.runLater(30000, () -> {
                    p.session.broadcast(new NetMessage.OnGameEnd(PlayerRole.Impostor));
                });
            }
        });
        handlers.register(NetMessage.FixSabotage.class, (p, m) -> {
            if (p.session == null) return;
            if (m.fixed) p.session.fixers++;
            else p.session.fixers--;

            if (m.sabotage == Sabotage.Reactor || m.sabotage == Sabotage.O2) {
                if (p.session.fixers == 2) {
                    scheduler.cancel(p.session.sabotageTask);
                    p.session.broadcast(new NetMessage.OnSabotageStateChanged(m.sabotage, false, 0, ""));
                }
            } else if (m.sabotage == Sabotage.Comms || m.sabotage == Sabotage.Lights) {
                if (p.session.fixers > 0) {
                    p.session.broadcast(new NetMessage.OnSabotageStateChanged(m.sabotage, false, 0, ""));
                }
            }
        });
    }

    private void tpAllToSpawn(ServerSession session) {
        var center = new Vector3f(28.09f, 0.0f, -22.46f);
        var radius = 1.7f;
        var angleOffset = (2.0f * Math.PI) / session.getPlayers().size();
        var angle = 0f;
        for (var player : session.getPlayers()) {
            var position = new Vector3f(
                    center.x + MathF.sin(angle) * radius,
                    center.y,
                    center.z + MathF.cos(angle) * radius
            );
            session.broadcast(new NetMessage.PositionChange(player.getId(), position, 0));
            angle += angleOffset;
        }
    }

    private void handleVotingEnd(ServerSession session) {
        if (session.votingTask != null)
            scheduler.cancel(session.votingTask);
        session.broadcast(new NetMessage.OnVoteResults(session.votes));

        scheduler.runLater(8000, () -> {
            // Determine who was ejected

            Map<Integer, Integer> voteCounts = new HashMap<>();
            if (session.votes.isEmpty()) {
                session.broadcast(new NetMessage.OnPlayerEjected(NetMessage.OnPlayerEjected.Result.Tie, 0));
                return;
            }

            for (var vote : session.votes.entrySet()) {
                var dstPlayerId = vote.getValue();
                if (!voteCounts.containsKey(dstPlayerId))
                    voteCounts.put(dstPlayerId, 1);
                else
                    voteCounts.put(dstPlayerId, voteCounts.get(dstPlayerId) + 1);
            }
            var sorted = voteCounts.entrySet()
                    .stream()
                    .sorted(Comparator.comparingInt(Map.Entry::getValue))
                    .collect(Collectors.toList());

            var highestVotes = sorted.get(0);
            var secondHighest = sorted.size() > 1 ? sorted.get(1) : null;

            if (secondHighest != null && highestVotes.getValue().equals(secondHighest.getValue())) {
                // Tie
                session.broadcast(new NetMessage.OnPlayerEjected(NetMessage.OnPlayerEjected.Result.Tie, 0));
            } else if (highestVotes.getKey() == Player.SKIP_PLAYER) {
                // Skip
                session.broadcast(new NetMessage.OnPlayerEjected(NetMessage.OnPlayerEjected.Result.Skip, 0));
            } else {
                // Eject
                int ejectedPlayer = highestVotes.getKey();
                session.getPlayer(ejectedPlayer).player.alive = false;
                session.broadcast(new NetMessage.Kill(true, ejectedPlayer));
                session.broadcast(new NetMessage.OnPlayerEjected(NetMessage.OnPlayerEjected.Result.Eject, ejectedPlayer));
            }

            tpAllToSpawn(session);

            session.votes.clear();

            checkVictory(session);
        });
    }

    private void checkAllVoted(ServerSession session) {
        if (session.votes.size() == session.getPlayers().size())
            handleVotingEnd(session);
    }

    private void checkVictory(ServerSession session) {
        var winners = findWinners(session);
        if (winners == null) return;
        session.broadcast(new NetMessage.OnGameEnd(winners));
    }

    private PlayerRole findWinners(ServerSession session) {
        if (session.getTaskProgress() == 1.0f)
            return PlayerRole.Crewmate;

        var impostors = 0;
        var crewmates = 0;
        for (var player : session.getPlayers()) {
            if (!player.player.alive) continue;
            if (player.getRole() == PlayerRole.Impostor) impostors++;
            else crewmates++;
        }
        if (impostors == 0) return PlayerRole.Crewmate;
        else if (impostors >= crewmates) return PlayerRole.Impostor;
        else return null;
    }

    private void generateTasks(List<PlayerTask> list, SessionConfig config) {
        for (var i = 0; i < config.getShortTasks(); i++) {
            PlayerTask shortTask = generate(list, TaskGenerator::newShortTask);
            list.add(shortTask);
        }

        for (var i = 0; i < config.getLongTasks(); i++) {
            PlayerTask longTask = generate(list, TaskGenerator::newLongTask);
            list.add(longTask);
        }
    }

    private <T> T generate(List<T> list, Supplier<T> supplier) {
        T t;
        var tries = 0;
        do {
            t = supplier.get();
            tries++;
            if (tries > 25)
                throw new IllegalStateException("Exceeded random tries.");
        } while (list.contains(t));
        return t;
    }

    private void broadcastPlayerUpdate(PlayerConnection p) {
        p.session.broadcast(new NetMessage.OnPlayerUpdate(
                p.player.id,
                p.player.color,
                p.player.role
        ));
    }

    private PlayerConnection randomImpostor(ServerSession session) {
        PlayerConnection player;
        do {
            player = RandomUtil.getRandomItem(session.getPlayers());
        } while (player.player.role == PlayerRole.Impostor); // Don't select impostors twice
        return player;
    }

    private int newId() {
        return idCounter.incrementAndGet();
    }

    private ServerSession getSession(String code) {
        return code == null || code.isEmpty() ? null : sessions.get(code);
    }

}

File: src/main/kotlin/de/twometer/amogus/server/AmongUsServer.kt
package de.twometer.amogus.server

import com.esotericsoftware.kryonet.Connection
import com.esotericsoftware.kryonet.Listener
import com.esotericsoftware.kryonet.Server
import de.twometer.amogus.concurrency.AsyncScheduler
import de.twometer.amogus.model.*
import de.twometer.amogus.net.*
import de.twometer.neko.util.MathF
import de.twometer.neko.util.MathF.PI
import mu.KotlinLogging
import org.joml.Vector3f
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

private val logger = KotlinLogging.logger {}

object AmongUsServer : Server() {

    private val sessions = ConcurrentHashMap<String, ServerSession>()
    private val playerIdCounter = AtomicInteger()
    private const val disableVictory = false
    private val scheduler = AsyncScheduler()

    fun main() {
        registerAllNetMessages(kryo)
        start()
        bind(32783)

        addListener(object : Listener() {
            override fun received(conn: Connection, msg: Any) {
                handle(conn as PlayerClient, msg)
            }

            override fun connected(conn: Connection) {
                logger.info { "Client connected from ${conn.remoteAddressTCP}" }
            }

            override fun disconnected(conn: Connection) {
                val client = conn as PlayerClient
                logger.info { "Player #${client.id} disconnected" }
                val session = client.session ?: return

                session.players.remove(client)
                if (session.players.size == 0) {
                    sessions.remove(session.code)
                    logger.debug { "Session ${session.code} is now empty and was destroyed" }
                    return
                }

                if (client.isHost) {
                    logger.debug { "Finding new host for session ${session.code}" }
                    val newHost = session.players.filter { !it.isHost }.shuffled().first()
                    logger.debug { "Selected host: ${newHost.id}" }
                    session.host = newHost.id
                    session.broadcast(OnSessionUpdate(session.code, session.host, session.config))
                }

                session.broadcast(OnPlayerLeave(client.id))

                checkVictory(session)
            }
        })

        logger.info { "Server started and bound" }

        scheduler.start()
    }

    override fun newConnection(): Connection {
        return PlayerClient()
    }

    // God function lets go
    private fun handle(client: PlayerClient, msg: Any) {
        when (msg) {
            is HandshakeRequest -> {
                client.player.id = playerIdCounter.incrementAndGet()
                client.sendTCP(HandshakeResponse(msg.version == 2, client.player.id))
                logger.info { "Player #${client.id} connected" }
            }
            is SessionCreateRequest -> {
                val code = CodeGenerator.newSessionCode()
                val session = ServerSession(code, client.id)
                sessions[code] = session
                client.sendTCP(SessionCreateResponse(true, code))
                logger.info { "Player ${client.id} created session ${session.code}" }
            }
            is SessionJoinRequest -> {
                val session = sessions[msg.code]
                if (session == null)
                    client.sendTCP(SessionJoinResponse(false, "Session does not exist"))
                else if (session.players.any { it.username.equals(msg.username, true) })
                    client.sendTCP(SessionJoinResponse(false, "Username is taken"))
                else if (msg.username.isBlank())
                    client.sendTCP(SessionJoinResponse(false, "Invalid username"))
                else if (session.isFull)
                    client.sendTCP(SessionJoinResponse(false, "Lobby full"))
                else {
                    logger.info { "Player ${client.id} joining session ${session.code}" }
                    client.session = session
                    client.player.username = msg.username
                    client.player.role = PlayerRole.Crewmate
                    client.player.color = session.randomColor()
                    client.sendTCP(SessionJoinResponse(true))
                    session.players.forEach {
                        client.sendTCP(OnPlayerJoin(it.id))
                        client.sendTCP(OnPlayerUpdate(it.id, it.username, it.color, it.role))
                    }
                    client.sendTCP(OnSessionUpdate(session.code, session.host, session.config))
                    session.players.add(client)
                    session.broadcast(OnPlayerJoin(client.id))
                    session.broadcast(OnPlayerUpdate(client.id, client.username, client.color, client.role))
                }
            }
            is SessionConfigureRequest -> {
                if (!client.isHost) return
                logger.info { "Player ${client.id} reconfiguring their session" }
                val session = client.session!!
                session.config = msg.config
                client.sendTCP(SessionConfigureResponse(true))
                session.broadcast(OnSessionUpdate(session.code, session.host, session.config))
            }
            is ColorChangeRequest -> {
                if (!client.inSession) return
                logger.info { "Player ${client.id} attempting to change their color from ${client.color} to ${msg.newColor}" }
                if (client.session!!.players.any { c -> c.color == msg.newColor })
                    client.sendTCP(ColorChangeResponse(false))
                else {
                    client.player.color = msg.newColor
                    client.session!!.broadcast(OnPlayerUpdate(client.id, client.username, client.color, client.role))
                    client.sendTCP(ColorChangeResponse(true))
                }
            }
            is CastVote -> {
                if (!client.inSession) return
                logger.info { "Player ${client.id} casting vote for ${msg.playerId}" }
                client.session!!.broadcast(OnPlayerVoted(client.id, msg.playerId))
                client.session!!.votes[client.id] = msg.playerId
            }
            is KillPlayer -> {
                if (!client.inSession) return
                if (client.role != PlayerRole.Impostor) return
                logger.info { "Impostor ${client.id} killing player ${msg.playerId}" }
                client.session!!.findPlayer(msg.playerId)?.run {
                    player.state = PlayerState.Ghost
                    client.session!!.broadcast(OnPlayerKilled(this.id))
                    checkVictory(client.session!!)
                }
            }
            is StartGame -> {
                if (!client.isHost) return
                logger.info { "Player ${client.id} starting their game" }
                val session = client.session!!
                // ReZero
                resetSession(session)
                // Select impostors
                repeat(session.config.impostorCount) {
                    selectImpostor(session)?.player?.role = PlayerRole.Impostor
                }
                // Gigasync
                session.broadcast(OnSessionUpdate(session.code, session.host, session.config))
                session.players.forEach {
                    session.broadcast(OnPlayerUpdate(it.id, it.username, it.color, it.role))
                }
                assignSpawnPositions(session)
                // And lezzgooooooooo :3
                val commonTask = TaskGenerator.newCommonTask()
                session.players.forEach {
                    val tasks = ArrayList<PlayerTask>()
                    tasks.add(commonTask)
                    repeat(session.config.shortTasks) { tasks.add(TaskGenerator.newShortTask()) }
                    repeat(session.config.longTasks) { tasks.add(TaskGenerator.newLongTask()) }
                    tasks.shuffle()
                    if (it.role != PlayerRole.Impostor) session.totalTaskStages += tasks.sumOf { task -> task.length }
                    it.sendTCP(OnGameStarted(tasks))
                }
            }
            is SabotageStart -> {
                val session = client.session ?: return
                if (client.role != PlayerRole.Impostor) return
                synchronized(session.sabotageLock) {
                    if (session.currentSabotage != null) return
                    session.currentSabotage = msg.sabotage
                    logger.info { "Impostor ${client.id} sabotaging ${msg.sabotage}" }
                    session.broadcast(OnSabotageChanged(msg.sabotage, CodeGenerator.newO2Code()))
                    if (msg.sabotage.critical) {
                        logger.debug { "Starting kill task" }
                        session.sabotageKillTask = scheduler.runLater(45000) {
                            logger.debug { "Sabotage to ${msg.sabotage} killed everyone" }

                            // Kill everyone!
                            session.players.forEach {
                                if (it.role == PlayerRole.Crewmate) it.player.state = PlayerState.Ghost
                            }

                            // And let the victory check do the rest
                            checkVictory(session)
                        }
                    }
                }
                logger.debug { "Sabotage started" }
            }
            is SabotageFix -> {
                val session = client.session ?: return
                logger.debug { "Sabotage fix request handler called" }
                synchronized(session.sabotageLock) {
                    val sabotage = session.currentSabotage ?: return
                    if (!sabotage.critical && msg.fixing) {
                        logger.debug { "Player ${client.id} fixed $sabotage" }
                        session.clearSabotage()
                    } else if (sabotage == SabotageType.O2) {
                        if (msg.fixing)
                            session.sabotageFixedLocations.add(msg.location)

                        logger.debug { "Player ${client.id} entered ${session.sabotageFixedLocations.size}/2 O2 codes" }

                        if (session.sabotageFixedLocations.size == 2) {
                            session.clearSabotage()
                        }
                    } else if (sabotage == SabotageType.Reactor) {
                        if (msg.fixing)
                            session.sabotageFixingPlayers++
                        else
                            session.sabotageFixingPlayers--

                        logger.debug { "${session.sabotageFixingPlayers}/2 players fixing reactor meltdown" }

                        if (session.sabotageFixingPlayers == 2) {
                            session.clearSabotage()
                        }
                    }
                }
            }
            is CompleteTaskStage -> {
                val session = client.session ?: return
                if (client.role != PlayerRole.Crewmate) return
                session.tasksCompleted++
                session.broadcast(OnTaskProgress(session.tasksCompleted / session.totalTaskStages.toFloat()))
                checkVictory(session)
            }
            is CallMeeting -> {
                val session = client.session ?: return
                logger.info { "Player ${client.id} called an emergency meeting (by_button = ${msg.byButton})" }
                session.broadcast(OnEmergencyMeeting(client.id, msg.byButton))
                session.votes.clear()
                assignSpawnPositions(session)
            }
            is ChangePosition -> {
                client.session?.broadcastExcept(OnPlayerMove(client.id, msg.pos, msg.rot), client.id)
            }
            is ChangeCameraState -> {
                val session = client.session ?: return
                synchronized(session.surveillanceLock) {
                    if (msg.inCams) session.surveillancePlayers++
                    else if (session.surveillancePlayers > 0) session.surveillancePlayers--
                    session.broadcast(OnSurveillanceChanged(session.surveillancePlayers > 0))
                }
            }
            is RequestEjectionResult -> {
                val session = client.session ?: return
                synchronized(session.votingTimerLock) {
                    session.votingTimerCompletePlayers++
                    if (session.votingTimerCompletePlayers >= session.players.size) {
                        val ejectionResult = session.determineEjection()
                        logger.info { "Session ${session.code} voted for $ejectionResult" }
                        session.broadcast(OnEjectionResult(ejectionResult))
                        session.votingTimerCompletePlayers = 0
                        if (ejectionResult.type == EjectResultType.Ejected)
                            session.findPlayer(ejectionResult.player)?.player?.state = PlayerState.Ghost
                        checkVictory(session)
                    }
                }
            }
        }
    }

    private fun assignSpawnPositions(session: ServerSession) {
        val center = Vector3f(28.09f, 0.0f, -22.46f)
        val radius = 2.25f
        val angleIncrement = 2.0f * PI / session.players.size
        var angle = 0f
        session.players.forEach {
            val pos = Vector3f(center).add(MathF.sin(angle) * radius, 0f, MathF.cos(angle) * radius)
            session.broadcast(OnPlayerMove(it.id, pos, 0f))
            angle += angleIncrement
        }
    }

    private fun resetSession(session: ServerSession) {
        session.votes.clear()
        session.tasksCompleted = 0
        session.totalTaskStages = 0
        session.surveillancePlayers = 0
        session.votingTimerCompletePlayers = 0
        session.clearSabotage()
        session.players.forEach {
            it.player.state = PlayerState.Alive
            it.player.role = PlayerRole.Crewmate
        }
    }

    private fun ServerSession.clearSabotage() {
        currentSabotage = null
        sabotageKillTask?.also { scheduler.cancel(it) }
        sabotageKillTask = null
        sabotageFixedLocations.clear()
        sabotageFixingPlayers = 0
        broadcast(OnSabotageChanged(null))
    }

    private fun selectImpostor(session: ServerSession): PlayerClient? =
        session.players.filter { it.role != PlayerRole.Impostor }.shuffled().firstOrNull()

    private fun checkVictory(session: ServerSession) {
        if (disableVictory) return
        val winners = tryFindWinners(session) ?: return
        session.broadcast(OnGameEnded(winners))
        resetSession(session)
    }

    private fun tryFindWinners(session: ServerSession): PlayerRole? {
        if (session.tasksCompleted >= session.totalTaskStages)
            return PlayerRole.Crewmate
        val alivePlayers = session.players.filter { it.state == PlayerState.Alive }
        val numAliveImpostors = alivePlayers.count { it.role == PlayerRole.Impostor }
        val numAliveCrewmates = alivePlayers.count { it.role == PlayerRole.Crewmate }
        return when {
            numAliveImpostors == 0 -> PlayerRole.Crewmate                   // Crewmates ejected all impostors off the ship
            numAliveImpostors >= numAliveCrewmates -> PlayerRole.Impostor   // Not enough crewmates to vote any impostor off -> loss
            else -> null                                                    // The game is still on.
        }
    }

    private fun ServerSession.determineEjection(): EjectResult {
        if (votes.isEmpty())
            return EjectResult(EjectResultType.Tie)

        // Create map of <PlayerId: Int, NumVotes: Int>
        val accumulatedVotes = HashMap<Int, Int>()
        votes.entries.forEach {
            val numVotes = accumulatedVotes[it.value]
            if (numVotes != null)
                accumulatedVotes[it.value] = numVotes + 1
            else
                accumulatedVotes[it.value] = 1
        }

        // Create list of MapEntry<PlayerId, NumVotes>, sorted by NumVotes
        val sortedVotes = accumulatedVotes.entries.sortedBy { it.value }

        // Get the top two entries
        val highestVotes = sortedVotes[0]
        val secondHighestVotes = sortedVotes.getOrNull(1)

        // Determine the ejection result
        return if (secondHighestVotes?.value == highestVotes.value) {
            EjectResult(EjectResultType.Tie)
        } else if (highestVotes.key == IPlayer.INVALID_PLAYER_ID) {
            EjectResult(EjectResultType.Skipped)
        } else {
            EjectResult(EjectResultType.Ejected, highestVotes.key)
        }
    }

}